### PR TITLE
Fix interp_points + add assertion in test

### LINKED ISSUE #69: test_interp does not contain any assert statements
under the CI testing approach, all tests need asserts so that they can be checked automatically - we are no longer relying on checking the outputs from print statements manually.

### LINKED ISSUE #100: Check interp_points with news coords() and related tests

### PR DESCRIPTION
**Pull request by rhugonnet**
Changed `interp_points()` to `scipy.ndimage.map_coordinates` instead of griddata. Commented the rest of the code that might allow to write a mutiprocessing loop if needed in the future.
Added assertions to `test_interp()`.
All tests passing.
Resolves #100
Resolves #69

One current limitation: we need to know if the pixel coordinates refers to center or corner (is_area or is_point). Right now I shift x/y by 0.5 pixel assuming it's the center, which works for our test Raster.
For robustness, we'd need to add the is_area or is_point info from GDAL into the Raster class. This is necessary to further improve the inter_points method.

### CODE REVIEW

**Comment by rhugonnet** [comment]
Resolves #100 and #69

File: geoutils/georaster.py
Comment (erikmannerfelt) [inline geoutils/georaster.py:19]: `scipy` is not in the `setup.py` requirements.

File: geoutils/georaster.py
Comment (erikmannerfelt) [inline geoutils/georaster.py:1278]: Have you tried using `np.vectorize` here?

Something like:
```python

ind_invalid = np.vectorize(lambda x, y: self.outside_image(x, y, index=True))(j, i)
```
Using `np.vectorize()` is often orders of magnitude faster than python loops.

File: geoutils/georaster.py
Comment (erikmannerfelt) [inline geoutils/georaster.py:1262]: Does this have an overhead?
An alternative may be:
```python
x, y = np.asarray(pts, dtype=float) # maybe .T has to be added. Not sure
```

**Review by erikmannerfelt — COMMENTED** [review]
I only have some performance suggestions.

File: geoutils/georaster.py
Comment (rhugonnet) [reply geoutils/georaster.py:1278]: Done!

File: geoutils/georaster.py
Comment (rhugonnet) [reply geoutils/georaster.py:19]: Done

File: geoutils/georaster.py
Comment (adehecq) [inline geoutils/georaster.py:1223]: It's strange to have the double name x and i or y and j. It's a bit redundant...

File: geoutils/georaster.py
Comment (adehecq) [inline geoutils/georaster.py:1239]: What exactly is the expected format? From l 1258 it seems like it is ( (x1, y1), (x2, y2)...)?
If so, mention that shape should be (N, 2).

File: geoutils/georaster.py
Comment (adehecq) [inline geoutils/georaster.py:1262]: If we don't, we should definitely have a small function for this type of operation in projtools !

File: geoutils/georaster.py
Comment (adehecq) [inline geoutils/georaster.py:1272]: what's the difference between np.array and np.asarray?

File: geoutils/georaster.py
Comment (adehecq) [inline geoutils/georaster.py:1279]: I think it would be better to implement this in xy2ij and use it.
You can find the info about center/corner in the dataset tags: `self.ds.tags()`

File: geoutils/georaster.py
Comment (adehecq) [inline geoutils/georaster.py:1285]: This makes the assumption that the full data has been loaded, in particular all bands...

File: tests/test_georaster.py
Comment (adehecq) [inline tests/test_georaster.py:216]: It would be good to check that it works fine after the data has been reprojected, cropped or downsampled...

File: geoutils/georaster.py
Comment (rhugonnet) [reply geoutils/georaster.py:1223]: This is because input can be coordinate or index

File: geoutils/georaster.py
Comment (rhugonnet) [reply geoutils/georaster.py:1272]: I think np.asarray is the official one for convert all types of Iterables: https://numpy.org/doc/stable/reference/generated/numpy.asarray.html
Don't really know the difference, maybe @erikmannerfelt can help here

Comment (erikmannerfelt) [reply geoutils/georaster.py:1272]: From the documentation linked by @rhugonnet:

> Array interpretation of a. No copy is performed if the input is already an ndarray with matching dtype and order. If a is a subclass of ndarray, a base class ndarray is returned.

It's a zero-copy operation, assuming the array is already an array. Otherwise it will convert to an array. Calling `np.array(arr)` makes for potential copying, which is unnecessary overhead.

File: geoutils/georaster.py
Comment (erikmannerfelt) [reply geoutils/georaster.py:1285]: Could 
```python
if not self.isLoaded:
    self.load(1)

```
be added above to fix this?

Speaking of camelCase, what's up with that variable??

File: geoutils/georaster.py
Comment (adehecq) [reply geoutils/georaster.py:1272]: Good to know!

File: geoutils/georaster.py
Comment (adehecq) [reply geoutils/georaster.py:1285]: That's not sufficient. If isLoaded is True, but only one band was loaded, it won't work...
That's one more argument to save the info of loaded band in an attribute...

Comment (rhugonnet) [reply geoutils/georaster.py:1285]: I guess I'll add to all the other "is_loaded" issues :D

File: geoutils/georaster.py
Comment (rhugonnet) [reply geoutils/georaster.py]: Done. `self.ds.tags()` was used already

File: geoutils/georaster.py
Comment (adehecq) [inline geoutils/georaster.py:1184]: area is actually upper left

File: geoutils/georaster.py
Comment (adehecq) [inline geoutils/georaster.py:1202]: It's a bit strange to have the area_or_point info but only use it in rare cases. Why not have an option 'area_or_point', which could be set to area, point or None. In the latter case, it would be set to the value in the metadata.
If you want to keep the same as current, you can just set default to 'area'.

File: tests/test_georaster.py
Comment (rhugonnet) [reply tests/test_georaster.py]: Done for crop. This also changes the GDAL interpretation to "Area" instead of "Point" for our test sample. So I tested that this worked correctly as well.

File: geoutils/georaster.py
Comment (rhugonnet) [reply geoutils/georaster.py:1184]: You mean upper left in y/x dimension? Because y is inverted? So in indexes: Area = (0,0) and Point = (0.5,0.5), right? The new tests work that way.

Comment (adehecq) [reply geoutils/georaster.py:1184]: Yes, y axis is downward, so (0,0) is the upper left corner.

File: tests/test_georaster.py
Comment (adehecq) [inline tests/test_georaster.py:221]: Why not have `r.ds.transform[0]*1.5` instead of `list(r.ds.transform)[0] + list(r.ds.transform)[0]/2`?

File: tests/test_georaster.py
Comment (adehecq) [inline tests/test_georaster.py:249]: It's weird that crop changes the area_or_point attribute...

Comment (rhugonnet) [reply tests/test_georaster.py:249]: You think it comes from `_update()` ?

Comment (adehecq) [reply tests/test_georaster.py:249]: You can test...

Comment (adehecq) [reply tests/test_georaster.py:249]: If you can't find out why, maybe add as an issue.

**Comment by rhugonnet** [comment]
No wonder I couldn't get my head around those tests. 
I've discovered a new interesting behaviour of the `rio.ds.index()` function:
```python
r.ds.index(x,y)
Out[33]: (75, 301)
r.ds.index(x,y,op=np.float32)
Out[34]: (75.0, 302.0)
```
Just by changing operator, you shift y by 1. Right now `test_interp()` works the way it is intended. But, when comparing to `value_at_coords()`, it only works if the operator is math.floor (lower index).

How should we go about this?

File: tests/test_georaster.py
Comment (rhugonnet) [reply tests/test_georaster.py]: Removed

File: tests/test_georaster.py
Comment (rhugonnet) [reply tests/test_georaster.py:251]: Added #127

**Comment by erikmannerfelt** [comment]
> No wonder I couldn't get my head around those tests.
> I've discovered a new interesting behaviour of the `rio.ds.index()` function:
> 
> ```python
> r.ds.index(x,y)
> Out[33]: (75, 301)
> r.ds.index(x,y,op=np.float32)
> Out[34]: (75.0, 302.0)
> ```
> 
> Just by changing operator, you shift y by 1. Right now `test_interp()` works the way it is intended with `np.float32`. But, when comparing to `value_at_coords()`, it only works if the operator is math.floor (lower y index).
> 
> How should we go about this?

Hahaha what!? That should probably be added as an issue on rasterio, no? A quick search in their issues shows no result corresponding to this phenomenon.

Could this be circumvented for now, and then hopefully be fixed by rasterio in time?

**Comment by rhugonnet** [comment]
Wait for the next level:
```python
r.ds.index(x,y,op=np.float32,precision=10)
Out[20]: (138.0, 504.0)
r.ds.index(x,y,op=np.float32)
Out[21]: (138.0, 504.0)
r.ds.index(x,y)
Out[22]: (138, 503)
r.ds.index(x,y,precision=0)
Out[23]: (137, 504)
r.ds.index(x,y,precision=1)
Out[24]: (137, 504)
r.ds.index(x,y,precision=10)
Out[25]: (138, 504)
```
The issue is that the "math.floor" argument is by default in the underlying function:
```python
def rowcol(transform, xs, ys, op=math.floor, precision=None):
```
which is wrapper by `index`

**Comment by rhugonnet** [comment]
Need to test this in more details, but it looks like the "right" solution is the `np.float32` and the default setting with `math.floor` can be falsely rounding (with default `precision=None` or other) when you provide the exact coordinates:
```python
r.bounds
Out[27]: BoundingBox(left=478000.0, bottom=3088490.0, right=502000.0, top=3108140.0)
xmin = 478000
yax=3108140
ymax=3108140
r.ds.index(xmin,ymax)
Out[31]: (0, 0)
r.ds.index(xmin,ymax,op=np.float32)
Out[32]: (0.0, 0.0)
r.ds.index(xmin,ymax,op=np.float32,precision=10)
Out[33]: (0.0, 3.637979e-12)
r.ds.index(xmin,ymax,precision=10)
Out[34]: (0, 0)
r.ds.index(xmin,ymax,precision=1)
Out[35]: (-1, 0)
r.ds.index(xmin,ymax,precision=0)
Out[36]: (-1, 0)
r.ds.index(xmin,ymax)
Out[37]: (0, 0)
```

**Comment by erikmannerfelt** [comment]
> Wait for the next level:
> 
> ```python
> r.ds.index(x,y,op=np.float32,precision=10)
> Out[20]: (138.0, 504.0)
> r.ds.index(x,y,op=np.float32)
> Out[21]: (138.0, 504.0)
> r.ds.index(x,y)
> Out[22]: (138, 503)
> r.ds.index(x,y,precision=0)
> Out[23]: (137, 504)
> r.ds.index(x,y,precision=1)
> Out[24]: (137, 504)
> r.ds.index(x,y,precision=10)
> Out[25]: (138, 504)
> ```
> 
> What's problematic is that the "math.floor" argument is by default in the underlying function, and is the one showing all those changes:
> 
> ```python
> def rowcol(transform, xs, ys, op=math.floor, precision=None):
> ```
> 
> which is wrapped by `index`

Call an exorcist!

This should definitely be an issue in rasterio, no? 

Also, why do they need `math.floor()` in the first place? Maybe it's new (and rio still supports py=2.7), but `round(1.1)` returns `1` (`type(round(1.1)) == int`).

**Comment by adehecq** [comment]
I thought we had discussed this before, but I cannot find any mention to it. But I agree, the use of floor by default is a bit strange. Maybe float32 should be the default. We could add an optional argument opt to xy2ij too.

**Comment by adehecq** [comment]
@rhugonnet what is the status of this??

**Comment by adehecq** [comment]
@rhugonnet  Tomorrow, your first challenge will be to merge this PR! :-P

File: geoutils/georaster.py
Comment (rhugonnet) [reply geoutils/georaster.py:1234]: fixed in 37da8c2

File: geoutils/georaster.py
Comment (rhugonnet) [reply geoutils/georaster.py]: Done in 
903c260

**Comment by rhugonnet** [comment]
@erikmannerfelt @adehecq 
Defaulted to np.float32 as operator. Raises warning when other operator is used.
Modified area_or_point as optional input (None, 'Area', 'Point') that defaults to GDAL metadata AREA_OR_POINT fetched through `rasterio` when the value is None.
Updated all tests, passing.

**Comment by erikmannerfelt** [comment]
There seems to be a merge conflict. Otherwise, it looks very good!

**Comment by erikmannerfelt** [comment]
There should be an `Iterable` import here:
https://github.com/GlacioHack/GeoUtils/blob/5e2b975a1a281fb38e1695e58916b4eadddeaeb1/geoutils/georaster.py#L9

**Comment by rhugonnet** [comment]
I've fixed some additional bugs after merging.
Now only need approval to merge the PR!

**Review by erikmannerfelt — APPROVED** [review]
Great job!!

**Pull request merged**
